File: 74. Search a 2D Matrix/main.cpp
#include <vector>

using namespace std;

class Solution {
public:
    bool searchMatrix(vector<vector<int>> &matrix, int target) {
        int rowSize = matrix.size();
        if (rowSize <= 0) {
            return false;
        }

        int colSize = matrix[0].size();
        if (colSize <= 0) {
            return false;
        }

        int left = 0;
        int right = rowSize * colSize - 1;

        while (left <= right) {
            int mid = (left + right) / 2;
            int row = mid / colSize;
            int col = mid % colSize;

            if (target < matrix[row][col]) {
                right = mid - 1;
            } else if (target > matrix[row][col]) {
                left = mid + 1;
            } else {
                return true;
            }
        }

        return false;
    }
};

File: 74. Search a 2D Matrix/main.rs
struct Solution;

impl Solution {
    pub fn search_matrix(matrix: Vec<Vec<i32>>, target: i32) -> bool {
        let row_size = matrix.len() as i32;
        if row_size <= 0 {
            return false;
        }

        let col_size = matrix[0].len() as i32;
        if col_size <= 0 {
            return false;
        }

        let mut left = 0;
        let mut right = row_size * col_size - 1;

        while left <= right {
            let mid = (left + right) / 2;
            let row = (mid / col_size) as usize;
            let col = (mid % col_size) as usize;

            if target < matrix[row][col] {
                right = mid - 1;
            } else if target > matrix[row][col] {
                left = mid + 1;
            } else {
                return true;
            }
        }

        false
    }
}
